src/map/homunculus.cpp

```diff
@@ -343,68 +330,40 @@ int hom_delete(struct homun_data *hd, int emote)
 }
 
 /**
-* Calculates homunculus skill tree
-* @param hd
-* @param flag_envolve
-*/
-void hom_calc_skilltree(struct homun_data *hd, bool flag_evolve) {
-	uint8 i;
-	short c = 0;
-
+ * Calculates homunculus skill tree for specific evolve/class.
+ * @param hd: Homunculus data
+ * @param skill_tree: Homunculus db skill tree
+ */
+void hom_calc_skilltree_sub(homun_data *hd, std::vector<s_homun_skill_tree_entry> skill_tree) {
 	nullpo_retv(hd);
 
-	/* load previous homunculus form skills first. */
-	if (hd->homunculus.prev_class != 0 && (c = hom_class2index(hd->homunculus.prev_class)) >= 0) {
-		for (i = 0; i < MAX_HOM_SKILL_TREE; i++) {
-			uint16 skill_id;
-			short idx = -1;
-			bool fail = false;
-			if (!(skill_id = hskill_tree[c][i].id) || (idx = hom_skill_get_index(skill_id)) == -1)
-				continue;
-			if (hd->homunculus.hskill[idx].id)
-				continue; //Skill already known.
-			if (!battle_config.skillfree) {
-				uint8 j;
-				if (hskill_tree[c][i].need_level > hd->homunculus.level)
-					continue;
-				for (j = 0; j < MAX_HOM_SKILL_REQUIRE; j++) {
-					if (hskill_tree[c][i].need[j].id &&
-						hom_checkskill(hd,hskill_tree[c][i].need[j].id) < hskill_tree[c][i].need[j].lv)
-					{
-						fail = true;
-						break;
-					}
-				}
-			}
-			if (!fail)
-				hd->homunculus.hskill[idx].id = skill_id;
-		}
-	}
+	bool evolved = false;
 
+	if (hd->homunculus.class_ == hd->homunculusDB->evo_class)
+		evolved = true;
 
-	if ((c = hom_class2index(hd->homunculus.class_)) < 0)
-		return;
+	for (const auto &skit : skill_tree) {
+		uint16 skill_id = skit.id;
+		short idx = hom_skill_get_index(skill_id);
 
-	for (i = 0; i < MAX_HOM_SKILL_TREE; i++) {
-		unsigned int intimacy = 0;
-		uint16 skill_id;
-		short idx = -1;
-		bool fail = false;
-		if (!(skill_id = hskill_tree[c][i].id) || (idx = hom_skill_get_index(skill_id)) == -1)
+		if (skill_id == 0 || idx == -1)
 			continue;
 		if (hd->homunculus.hskill[idx].id)
 			continue; //Skill already known.
-		intimacy = (flag_evolve) ? 10 : hd->homunculus.intimacy;
-		if (intimacy < hskill_tree[c][i].intimacy * 100)
+		if (skit.intimacy > 0 && hd->homunculus.intimacy < skit.intimacy * 100) {
 			continue;
+		}
+		if (skit.evolution && !evolved) {
+			continue;
+		}
+
+		bool fail = false;
+
 		if (!battle_config.skillfree) {
```

Comment:
Shouldnt this config also apply for intimacy and evolution?
Guess then the whole loop could be skipped maybe?